core/src/main/java/org/apache/calcite/plan/RelOptForeignKey.java

```diff
@@ -0,0 +1,411 @@
+ * correspond to each other. The order of elements in {@code uniqueColumns}
+ * and {@code foreignColumns} is consistent with the order of constraints in
+ * the constraints list.
+ *
+ * <p>For instance,
+ * <blockquote>
+ * <pre>select e.deptno, e.ename, d.deptno
+ * from emp as e
+ * inner join dept as d
+ * on e.deptno = d.deptno</pre></blockquote>
+ *
+ * <p>the foreign key is the DEPTNO column of CATALOG.SALES.EMP table,
+ * reference the DEPTNO unique column of CATALOG.SALES.DEPT table.
+ *
+ * <p>Invoke the {@link RelMetadataQuery#getConfirmedForeignKeys} method which
+ * input param is the top {@link org.apache.calcite.rel.core.Project},
+ * the following results can be obtained.
+ *
+ * <p>{@code constraints} is
+ * [{left: [CATALOG, SALES, EMP].$7.#true, right: [CATALOG, SALES, DEPT].$0.#true}]
+ * {@code foreignColumns} is {0}
+ * {@code uniqueColumns} is {2}
+ *
+ * <p>For instance,
+ * <blockquote>
+ * <pre>select name, deptno
+ * from dept</pre></blockquote>
+ *
+ * <p>{@code constraints} is
+ * [{left: null.null.#false, right: [CATALOG, SALES, DEPT].$0.#false}]
+ * {@code foreignColumns} is {}
+ * {@code uniqueColumns} is {1}
+ *
+ * <p>For instance,
+ * <blockquote>
+ * <pre>select ename, sal, deptno
+ * from emp</pre></blockquote>
+ *
+ * <p>{@code constraints} is
+ * [{left: [CATALOG, SALES, EMP].$7.#false, right: [CATALOG, SALES, DEPT].$0.#false}]
+ * {@code foreignColumns} is {2}
+ * {@code uniqueColumns} is {}
+ *
+ * @see InferredConstraintKey
+ * @see org.apache.calcite.plan.RelOptForeignKey
+ */
+public class RelOptForeignKey {
+
+  /** Foreign key and unique key relationships in bottom-up derivation. */
+  private final List<Pair<InferredConstraintKey, InferredConstraintKey>> constraints;
+  /** Position of the foreign key on the current {@link org.apache.calcite.rel.RelNode}. */
+  private final ImmutableBitSet foreignColumns;
+  /** Position of the unique key on the current {@link org.apache.calcite.rel.RelNode}. */
+  private final ImmutableBitSet uniqueColumns;
+
+  private RelOptForeignKey(
+      List<Pair<InferredConstraintKey, InferredConstraintKey>> constraints,
+      ImmutableBitSet foreignColumns,
+      ImmutableBitSet uniqueColumns) {
+    this.constraints = constraints;
+    this.foreignColumns = foreignColumns;
+    this.uniqueColumns = uniqueColumns;
+  }
+
+  public static RelOptForeignKey of(List<Pair<InferredConstraintKey,
+      InferredConstraintKey>> constraints,
+      ImmutableBitSet foreignColumns,
+      ImmutableBitSet uniqueColumns) {
+    return new RelOptForeignKey(constraints, foreignColumns, uniqueColumns);
+  }
+
+  public ImmutableBitSet getForeignColumns() {
+    return foreignColumns;
+  }
+
+  public ImmutableBitSet getUniqueColumns() {
+    return uniqueColumns;
+  }
+
+  public List<Pair<InferredConstraintKey, InferredConstraintKey>> getConstraints() {
+    return constraints;
+  }
+
+  /** Returns the left side of a list of constraints. */
```

Comment:
What is meaning of left side,I think It should be described with clearer comments